Tut baroclinic gyre

jrscott opened the pull request:
## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)
docs: tutorial_baroclinic_gyre first commit, minor changes in tutorial_barotropic_gyre

jm-c commented:
@jrscott 
There are several levels of changes here that could have been put in separated commits (or even PR):
1) the split doc/examples/examples.rst ? dos this overlap with PR #245 ?
2) how P-E is written in algorithm/ section ?
and might be other that I did not see. The point is that it would make it easier to see what is changed.
Also, regarding eedata.mth, did you try to run the multi-threaded testreport both with and without MPI ? does not seem right to me.

jrscott commented:
@jm-c point 1: closed PR #245 
@jm-c point 2: indeed there are a few things that would be better in separate commits.
 - not sure this registers as much of a  "change" but I used the latex cal font for forcing terms, as this was how done in past tutorials -- but made section 2 consistent with this notation
(since this is related, don't think should be separate PR however)
 - I increased the toctree number depth to 5 THROUGHOUT the doc. Didn't notice any problems from this, but should have been flagged separately as a point to discuss
- also changed some stuff in tut BT gyre, e.g. moved discussion of phiHyd files to BC gyre
- split the configuration sections into "compile-time" and "run-time" to avoid any possible reader confusion.

jrscott commented:
@jm-c for this new setup, testreport runs ok single proc, MPI, and mth. Running MPI and mth combined is not documented anywhere; are you doing this in some testreport setups?

edoddridge commented:
I disagree with your logic on this one. The default behaviour of `pkg/mnc` is not an obscure detail. It's an important aspect of dealing with the output from a real simulation. Plus, sending a complete beginner to comment out two random lines in one data file is pretty obscure.

Other than modifying the `gluemncbig` call, are there other complications? NetCDF files are still broken up by tile, regardless of whether they get dumped in the top level directory or not.

Unless there's something I'm missing, my vote is strongly on the side of "leave the config alone and add a tiny bit of complexity to one analysis command".

jm-c commented:
May I suggest to meet in person so that we can clarify the choices here
and decide what is best ?

On Thu, Aug 29, 2019 at 08:35:41AM -0700, Ed Doddridge wrote:
> I disagree with your logic on this one. The default behaviour of `pkg/mnc` is not an obscure detail. It's an important aspect of dealing with the output from a real simulation. Plus, sending a complete beginner to comment out two random lines in one data file is pretty obscure.
> 
> Other than modifying the `gluemncbig` call, are there other complications? NetCDF files are still broken up by tile, regardless of whether they get dumped in the top level directory or not.
> 
> Unless there's something I'm missing, my vote is strongly on the side of "leave the config alone and add a tiny bit of complexity to one analysis command".
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/267#issuecomment-526241438

jrscott commented:
sure... this is a pretty minor issue anyway. @jm-c there is a lot more in this tutorial where I expect you will have more significant comments, questions, revisions etc.

jrscott commented:
thanks @edoddridge . That's fine, the main point here is showing how to plot and manipulate the data, do calculations etc. none of which requires making the plots pretty.

jrscott commented:
It would be nice to get this PR done soon; I believe the main thing we are waiting for is a thorough review by @jm-c (anticipating some minor changes).

On the mnc directory issue, let me re-try to explain why I think it is best to leave it as I have done.
Two things @edoddridge  wrote are not correct. First, MITgcm default behavior is to dump the mnc files in the top directory, same as mdsio. So for simplicity, I'm suggesting we do the same when mnc is first introduced, rather than further complicate things by explaining the non-default options and how they work (which was not immediately obvious to me). This tutorial is already too long and does too much :( Now of course the mnc file overwriting behavior is different, which I do state. The fact that testreport requires output in a subdirectory for mnc, that’s fine, but that’s basically irrelevant for tutorial teaching purposes. So the second point of @edoddridge  that I’d argue is off-base is about commenting out lines to run the tutorial: we’re already doing so in all other tutorials, so doing it here because of testreport is simply par for the course.  Moreover, if you did want to run the tutorial to dump files in a sub-directory, are you suggesting using the same name for testreport “mnc_test”?  That would be unfortunate. Another factor is that in the tutorial I carefully go through all output (including some binary output) and state what it is; it’s a lot easier on the user if all these files are in the top directory and they can stare at them (all in one place) when I go though this, assuming they have run this experiment on their own to peruse. I think if one weighs these considerations, it seems a pretty clear choice...

jm-c commented:
@jrscott  I took a quick look at this tutorial doc and, apart form the mnc_use_outdir story (would be easier to be discussed in person), I have few minor comments that I am listing here. But in general, it looks good.
1) in data.diagnostics I already fixed the &DIAG_STATIS_PARMS section:
   stat_fields(1:9,1) = 
has been changed to:
   stat_fields(1:2,1) =
 but this need to be changed also in corresponding docs section.

2) Open discussion: BC for theta: 
  => Yes, "zero-flux" is the right BC (except at the surface ! LinFS does imply
 a smal advective flux through model top, at z==0); however, we do not inforce
 zero gradient normal to BC, independently of diffKh & diffKr

3) 4.2.2 eta is not "diagnosed" ; replace the last 2 sentences of ths pragraph with:
 "The see surface height is found by solving implicitly the 2-D (eliptic) 
 surface pressure equation (see section 2.4).

4) 4.2.2.1, discussion about "Sa" : the factor 2 AND u ~ 1.m/s applies to HORIZONTAL
 advection and not vertcal advection
 I would suggest to replace "|u|" with "|c_max|" ; with c_max = maximum of horizontal
 flow speed (~ 1.m/s) or/and max of internal wave speed (< 3.m/s).

5) 4.2.3.1.1, simplify:
 gfd is a predefined "pacakage group" (see ...) of standard packages needed in most
 set-up; it is also the default compiled pkgs (..) settings and the minimum set 
 of pkgs necessary for GFD type setups. 
-> then continue with:  "In addition to ..."

6) again, why the quoting of "SIZE.h" has so many blank lines (from line 50 to line 64)
 at the end ? or maybe these blank lines only show up in my own html built ?

7) third bullet: suggest to described right away lines 41 & 42 of file "SIZE.h"
 Nx = sNx*nSx*nPx
 Ny = sNy*nSy*nPy
  this will make it much more clear why, when nPx=nPy=1, we need nSx=nSy=2
  since we need Nx=Ny=62 

8) 4.2.3.2.1, 
 a) again, BC on theta, not sure
 b) ivdc_kappa description: better to start with what is used (here: ivdc_kappa)
 rather than what is not used ("traditional convective adjustment").
 c) might has been more clear to rename file: 
    "Trest.bin" to "SST_relax.bin"
   since the first one might be confused with a 3-D initial temperature at "rest"

9) 4.2.3.2.2,
 I suggest to change the last sentence to:
 Apart these 2 additional packages, only standard ..

That's all.

jrscott commented:
@jm-c:

1. change pulled automatically into doc
2. Let's discuss if/how we can clarify in the doc
3. done. 
4.  done. Although |u| still in similar spot in BT gyre, let's discuss.
5. done, as specified. concern here is explicitness of "is the file packages.conf required?" which remains vague.
6. There was one extra blank line in SIZE.h, at the end (now removed), but my build is not showing this problem.
7. Let's discuss, confused about "third bullet". Correct though this is a important omission. Thinking maybe start with bullet showing lines 41-42+53-54, make your point in text here.
8. a-discuss; b, done, maybe add a ref to traditional approach?; c done
9. done

jm-c commented:
@jrscott  few other comments regarding the doc part:

1) section 4.2.3.2.1, PARM04 section:
replace: 
 Setting the origin in latitude is critical given that it affects the initialization of the Coriolis force
with:
 Setting the origin in latitude is critical given that it affects the setting of Coriolis parameter "f"
 (eq. 4.11 & 4.12).

2) section 4.2.4.1, Hydrostatic pressure description, problems with this sentence:

 Note that the linearized free surface contains water of density ρc and that its contribution to ϕ′ 
 is the from the full weight of the water in the free surface, (i.e., no reference density is subtracted), 
 in contrast with contributions below the free surface.
a) second line, what does the "from" means ?
b) not super clear, + the last part should be below z=0 and not below the free surface. 

I would suggest to change it to:
 Note that, with the linear free surface approximation,
 the contribution of the free surface position $\eta$ to ϕ′ involves the constant density $\rho_c$
 and not the density anomaly $\rho'$, in contrast with contributions from below $z = 0$.

3) section 4.2.5, after gnemake2 command:
change:
 Note we have added the option -mpi to the line which generates the makefile.
to:
 Note we have added the option -mpi to the genmake2 command that generates the makefile.

4) section 4.2.7 <- stop here but:
a) strange wording
  (but hey, in addition to the higher frequency of output without taxing our storage requirements, 
   isn’t it simpler to let MITgcm do the statistical number crunching than to write matlab or python code?)
I would remove this and also sentence just before.

b) fig.4.7: I would have reverse the color scale, so that: 
  Red <-> surface heating and Blue <-> surface cooling

c) could change u_bt and v_bt to upper case: U_bt, V_bt, just to make it more clean that they
  have different units than u,v ; also mention the units: m^2/s, the first time they are mentioned.

jrscott commented:
in that last (gluemncbig block) commit: might be opportunity to mention pip install mitgcmutils, thoughts? (I've kept it self-contained at the moment)

jrscott commented:
I think this is ready to be merged, unless folks want to do further review and suggest additional changes.

I have not done anything about the mnc issue. Here is the how the (non-default) mnc outdir option works:

Single processor (4 tiles):
mnc_0001/ - run #1, contains all four tiles 
mnc_0002/ - run #2, “                              “

MPI (4 cores):
mnc_0001 - process 1, run #1, tile 1 only
mnc_0002 - process 2, run #1, tile 2 only
...
mnc_0005 - process 1, run#2, tile 1 only
mnc_0006 - process 2, run#2, tile 2 only
...

If it were just the single processor behavior, I think there is a stronger argument here.
For the mpi run, gluemncbig on mnc_* files won’t even work if there are multiple runs, you need
to delete the earlier runs (which defeats the whole point!).

So in summary:

Doc as is: uses mnc default behavior, fairly consistent w/mdsio behavior except for that overwriting mnc output not allowed (mentioned in text). 
Requires user comment out lines in data.mnc (which they do also in data, settings for testreport). VERY SIMPLE!

Proposed change: have the user write output in directory mnc_test (i.e. borrowing the testreport subdir name, this is quite tragic but nobody had a better suggestion) -- which is NOT the mnc default behavior, and different than binary. Requires additional paragraphs explaining the above confusing rules for output subdir naming/contents. Should provide different gluemncbig syntax for single-process vs. mpi. And explain that the mpi gluemncbig syntax won’t work if you re-run and don’t delete previous output. Given that this tutorial is oriented to very beginner users, and this tutorial is already long and perhaps overwhelming, opting for this change seems rather ill-advised. It would be nice to document this somewhere however, certainly in the packages doc!

jrscott commented:
Update: @jahn looked further into gluemncbig and it seems it is smart enough to grab the LAST run data, however could get "confused" in some situations with different number of iterations being dumped etc.

jrscott commented:
any chance we can merge this now and start a new PR with additional edits?

edoddridge commented:
As discussed offline, I've modified the `pkg/mnc` behaviour to use multiple directories.

jm-c commented:
instead of 'label "documentation" ' I meant 'label "manual"'

edoddridge commented:
@jm-c - I've added a description to tag-index. If you're happy with it, then I reckon you should push the button. Probably squash merge this one.

jrscott commented:
yes, let's squash-merge this, then I can quickly make a new PR with converted (old) tutorials, then
I can do a new PRs for edits/fix up what is done so far.
